packages/interactivity/src/utils.ts

```diff
@@ -374,3 +374,23 @@ export const isPlainObject = (
 			typeof candidate === 'object' &&
 			candidate.constructor === Object
 	);
+
+/**
+ * Indicates that the passed `callback` requires synchronous access to the event object.
+ *
+ * @param callback The event callback.
+ * @return Wrapped event callback.
+ */
+export const withSyncEvent = ( callback: Function ): Function => {
+	if ( callback?.constructor?.name === 'GeneratorFunction' ) {
+		const wrapped = function* ( ...args: any[] ) {
+			yield* callback( ...args );
+		};
+		wrapped.sync = true;
+		return wrapped;
+	}
+
+	const wrapped = ( ...args: any[] ) => callback( ...args );
+	wrapped.sync = true;
+	return wrapped;
```

Comment:
Removing a bit of code duplication:

```suggestion
	let wrapped;
	if ( callback?.constructor?.name === 'GeneratorFunction' ) {
		wrapped = function* ( ...args: any[] ) {
			yield* callback( ...args );
		};
	} else {
		wrapped = ( ...args: any[] ) => callback( ...args );
	}
	wrapped.sync = true;
	return wrapped;
```

Reply:
I had tried to do this initially in the same way, but it doesn't work:
* The `let wrapped;` will require a type.
* Setting it to `Function` will cause the `.sync` assignment to not work because that's not available on a function object.
* Setting it to `any` would work, but not be great because `any`.

I just searched a bit further and found https://stackoverflow.com/questions/12766528/build-a-function-object-with-properties-in-typescript. That seems to work better, so I'm gonna go with that.

Reply:
Done as part of https://github.com/WordPress/gutenberg/pull/68097/commits/bf3dbcc861367a38131ac0943ad49f60e2474912.

I also identified a bug where the additional `sync` property was not being passed through, due to passing all store actions through the `withScope()` function, which happens via the `proxifyStore()` function.

By introducing an extra type `SyncAwareFunction`, this can be handled a bit more elegantly, as it allows to simply cast regular functions into that type and then set the `sync` property as needed (in practice, it's always `true`, or it's not set, i.e. a regular function).

I have tested this approach and it works well. It also ensures the original underlying JS types (e.g. `Function` or `GeneratorFunction`) are maintained.